### PR TITLE
Migrate to Typescript and Vue 3

### PR DESCRIPTION
**Pull request by mtklass**
This PR changes the processing code over to Typescript, which makes the code safer and faster (15% faster!). The modifications are meant to be minimal in terms of logic changes, and the site data should work as it did before.

The PR also migrates from Vue 2 to Vue 3, which was a harder change. It took a bit of fiddling to get everything working like before, but it should now behave just like the current twotech site.

### CODE REVIEW

**Comment by mtklass** [comment]
PR commits squashed.

File: src/components/BiomeList.vue
Comment (connorhsm) [inline src/components/BiomeList.vue:1]: Missing css for highlighted biome
master:
image

This PR:
image

File: src/components/BiomeInspector.vue [hunk 1]
```diff
@@ -1,81 +1,110 @@
...
 </script>
 
 <style lang="scss">
 .biomeInspector {
+  .loading {
+    text-align: center;
+    padding: 20px;
+    font-size: 18px;
+    color: #aaa;
+  }
```
Comment (connorhsm) [inline src/components/BiomeInspector.vue:107]: What does this target? Unable to find it

File: src/components/RecipeForLetters.vue
Comment (connorhsm) [inline src/components/RecipeForLetters.vue:1]: This is broken, though I forgot it existed, and I don't think I've ever seen someone use it.
image

Comment (connorhsm) [reply src/components/RecipeForLetters.vue:1]: As well as the link to get to it, which is in `ObjectInspector.vue`
image

**Review by connorhsm — APPROVED** [review]
Gold star for you, sir! 🌟

Appreciate all the effort that has gone into this, hopefully will make for a greater future!

Everything I have noted is mostly a nitpick and none of it blocks me from merging, so I am happy to merge this with your acknowledgement, and we can move these items for resolution to a single issue.

Search bar text has shadow, couldn't spot where this was happening specifically.
image

File: process/src/MainProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+    if (version) {
+      console.log("\nGenerating sitemap...");
+      console.time("Generating sitemap took");
+      gameData.generateSitemap();
+      console.timeEnd("Generating sitemap took");
+      return null;
+    }
+
+    console.timeEnd("Processing took");
```
Comment (connorhsm) [inline process/src/MainProcessor.ts:155]: Need to copy line 155 to under line 151, to catch the timeEnd for when we do process a versioned data set.

**Comment by mtklass** [comment]
Thanks! And good catches! Sounds like a plan to put them into an issue. Most shouldn't be too bad to tackle.

I'm good with merging this. 👍

File: process/src/MainProcessor.ts (hunk 2 above)
Comment (mtklass) [reply process/src/MainProcessor.ts:155]: Whoops, missed that! That'll be an easy one to tackle in the new issue.

File: src/components/RecipeForLetters.vue
Comment (mtklass) [reply src/components/RecipeForLetters.vue:1]: Ah, yeah I did have that working at one point, but I must have re-broken it somehow. I'll guess it to be a medium difficulty compared to other issues here.

File: src/components/BiomeInspector.vue (hunk 1 above)
Comment (mtklass) [reply src/components/BiomeInspector.vue:107]: Ah, I think it's vestigial, from trying to get something working earlier on, and it never got removed.

It's okay to remove this.

File: src/components/BiomeList.vue
Comment (mtklass) [reply src/components/BiomeList.vue:1]: Oh, yeah, missed that. Hopefully shouldn't be too hard to figure out.

File: process/src/GameObject.ts [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+  }
+
+  // parseValue(value) {
+  //   if (isNaN(value))
+  //     return value;
+  //   if (value.includes("."))
+  //     return parseFloat(value);
+  //   return parseInt(value);
+  // }
+
```
[the remark this answers is not in the document]
Comment (mtklass) [reply process/src/GameObject.ts:209]: Yep, leftover from before the migration. Will remove

File: process/src/GameData.ts [hunk 4]
```diff
@@ -102,45 +115,50 @@ class GameData {
...
   }
 
-  generateTechTree() {
-    var generator = new TechTreeGenerator();
-    generator.generate(Object.values(this.objects));
-  }
+  // generateTechTree() {
+  //   var generator = new TechTreeGenerator();
+  //   generator.generate(Object.values(this.objects));
+  // }
```
[the remark this answers is not in the document]
Comment (mtklass) [reply process/src/GameData.ts:131]: I couldn't find any reference to it in the older code either. I meant to just get rid of it at the end, but forgot. Let's drop it.

File: process/src/ChangeLogCommit.ts [hunk 5]
```diff
@@ -196,16 +232,29 @@ class ChangeLogCommit {
...
-module.exports = ChangeLogCommit;
+interface ExportedChangeLogCommitData {
+  sha: string;
+  message: string;
+  date: Date;
+  addedObjectIDs?: string[];
+  removedObjectIDs?: string[];
+  // TODO: Change this to ExportedTransitionData when it gets defined.
+  addedTransitions?: ExportedTransitionData[];
+  removedTransitions?: ExportedTransitionData[];
```
[the remark this answers is not in the document]
Comment (mtklass) [reply process/src/ChangeLogCommit.ts:255]: Good call, we should remove that comment.

**Comment by connorhsm** [comment]
For reference, this PR also introduced processing of specific objects/sprites. #44

**Pull request merged**